src/app/shared/store/registration.state.ts

```
@@ -184,41 +177,59 @@ export class RegistrationState {
-      case Role.providerDeputy:
-      case Role.employee:
-        // TODO: Add provider patch when profile/provider id will be provided
-        return this.employeeService.getEmployeeById(state.user.id).pipe(tap((employee: Employee) => patchState({ employee })));
-      case Role.ministryAdmin:
-        return this.ministryAdminService.getAdminProfile().pipe(tap((ministryAdmin: MinistryAdmin) => patchState({ ministryAdmin })));
-      case Role.regionAdmin:
-        return this.regionAdminService.getAdminProfile().pipe(tap((regionAdmin: RegionAdmin) => patchState({ regionAdmin })));
-      case Role.areaAdmin:
-        return this.areaAdmin.getAdminProfile().pipe(tap((areaAdmin: AreaAdmin) => patchState({ areaAdmin })));
+    const role = state.user.role as Role;
+    const profileKey = this.userProfileService.getStateKeyByRole(role);
+
+    patchState({ isAuthorizationLoading: true, role });
+
+    const cachedProfile = this.store.selectSnapshot(AppState.profile);
+
+    // if there is cached user profile
+    if (cachedProfile) {
+      patchState({ [profileKey]: cachedProfile, isAuthorizationLoading: false });
+      return;
+    }
+
+    const profileObservable = this.userProfileService.getProfileObservableByRole(role, state.user.id);
+    if (!profileObservable) {
+      patchState({ isAuthorizationLoading: false });
+      return;
     }
+
+    return profileObservable.pipe(
+      tap((profile: Parent | Provider | Employee | MinistryAdmin | RegionAdmin | AreaAdmin) => {
+        dispatch(new SetProfile(profile));
+        patchState({ [profileKey]: profile });
+      }),
+      catchError((error: HttpErrorResponse) => {
+        if (error.status === HttpStatusCode.Unauthorized || error.status === HttpStatusCode.Forbidden) {
+          this.router.navigate(['/forbidden']);
+        }
+        return throwError(() => new Error('Something went wrong!'));
```

it may be better to throw the original error